File: app/models/message.rb
class Message < FirestoreModel
  attribute :conversation_id, :string
  attribute :user_id, :string
  attribute :sender_type, :string
  attribute :content, :string
  attribute :message_type, :string, default: 'text'
  attribute :read, :boolean, default: false
  attribute :metadata, default: {}

  validates :sender_type, presence: true, inclusion: { in: %w[user reed system] }
  validates :content, presence: true
  validates :message_type, presence: true, inclusion: { in: %w[text audio image system] }

  after_create :update_conversation_timestamp

  def self.unread
    where(:read, false).get
  end

  def self.read
    where(:read, true).get
  end

  def self.recent
    order(:created_at, :desc).get
  end

  def mark_as_read!
    update(read: true)
  end

  def conversation
    Conversation.find(conversation_id) if conversation_id.present?
  end

  def user
    User.find(user_id) if user_id.present?
  end

  private

  def update_conversation_timestamp
    conv = conversation
    conv&.update_last_message!
  end
end

File: frontend/lib/models/message.dart
class Message {
  final int id;
  final int conversationId;
  final String senderType; // user, reed, system
  final String content;
  final String messageType; // text, audio, image, system
  final bool read;
  final DateTime createdAt;
  final Map<String, dynamic>? metadata;

  Message({
    required this.id,
    required this.conversationId,
    required this.senderType,
    required this.content,
    required this.messageType,
    required this.read,
    required this.createdAt,
    this.metadata,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] as int,
      conversationId: json['conversation_id'] as int,
      senderType: json['sender_type'] as String,
      content: json['content'] as String,
      messageType: json['message_type'] as String,
      read: json['read'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      metadata: json['metadata'] as Map<String, dynamic>?,
    );
  }

  bool get isFromUser => senderType == 'user';
  bool get isFromReed => senderType == 'reed';
  bool get isSystem => senderType == 'system';
}

